app/lib/package/export_api_to_bucket.dart

```diff
@@ -0,0 +1,297 @@
+  /// updates.
+  @visibleForTesting
+  Future<void> fullPkgScanAndUpload() async {
+    final pool = Pool(_concurrency);
+    final futures = <Future>[];
+    await for (final mp in dbService.query<ModeratedPackage>().run()) {
+      final f = pool.withResource(() => _deletePackageFromBucket(mp.name!));
+      futures.add(f);
+    }
+    await Future.wait(futures);
+    futures.clear();
+
+    await for (final package in dbService.query<Package>().run()) {
+      final f = pool.withResource(() async {
+        if (package.isVisible) {
+          await _uploadPackageToBucket(package.name!);
+        } else {
+          await _deletePackageFromBucket(package.name!);
+        }
+      });
+      futures.add(f);
+    }
+    await Future.wait(futures);
+    await pool.close();
+  }
+
+  @visibleForTesting
+  Future<void> incrementalPkgScanAndUpload(
+    GlobalLockClaim claim, {
+    Duration sleepDuration = const Duration(minutes: 2),
+  }) async {
+    final pool = Pool(_concurrency);
+    // The claim will be released after a day, another process may
+    // start to upload the API JSONs from scratch again.
+    final workUntil = clock.now().add(Duration(days: 1));
+
+    // start monitoring with a window of 7 days lookback
+    var lastQueryStarted = clock.now().subtract(Duration(days: 7));
+    while (claim.valid) {
+      final now = clock.now().toUtc();
+      if (now.isAfter(workUntil)) {
+        break;
+      }
+
+      // clear old entries from last seen cache
+      _pkgLastUpdated.removeWhere((key, event) =>
+          now.difference(event.updated) > const Duration(hours: 1));
+
+      lastQueryStarted = now;
+      final futures = <Future>[];
+      final eventsSince = lastQueryStarted.subtract(Duration(minutes: 5));
+      await for (final event in _queryRecentPkgUpdatedEvents(eventsSince)) {
+        if (!claim.valid) {
+          break;
+        }
+        final f = pool.withResource(() async {
+          if (!claim.valid) {
+            return;
+          }
+          final last = _pkgLastUpdated[event.package];
+          if (last != null && last.updated.isAtOrAfter(event.updated)) {
+            return;
+          }
+          _pkgLastUpdated[event.package] = event;
+          if (event.isVisible) {
+            await _uploadPackageToBucket(event.package);
+          } else {
+            await _deletePackageFromBucket(event.package);
+          }
+        });
+        futures.add(f);
+      }
+      await Future.wait(futures);
+      futures.clear();
+      await Future.delayed(sleepDuration);
+    }
+    await pool.close();
+  }
+
+  /// Uploads the package version API response bytes to the bucket, mirroring
+  /// the endpoint name in the file location.
+  Future<void> _uploadPackageToBucket(String package) async {
+    final data = await retry(() => packageBackend.listVersions(package));
+    final rawBytes = jsonUtf8Encoder.convert(data.toJson());
+    final gzippedBytes = gzip.encode(rawBytes);
+
```

Comment:
Why not compute md5 here?

Reply:
We do that in `_isSameContent`. Oh, we may do it twice, that's true...

Reply:
Added lazy calculation.